Add the script tag after-render

stefanpenner opened the pull request:
Kinda strange to add to the DOM during just any period, we likely want this to always happen when stuff has settled down. 

In-theory this could be made pluggable, allowing an application to provide an alternative schedular.

```js
import { setScriptScheduler } from '...';

setScriptScheduler(function(cb) {
  schedular.scheduleWork('afterContentPaint', cb);
});
```

cc @scalvert

stefanpenner committed:
Add the script tag after-render

trentmwillis commented on addon/loaders/js.js:
```diff
@@ -15,11 +15,17 @@ export default nodeLoader(function js(uri) {
-    document.head.appendChild(script);
+        document.head.appendChild(script);
+      } catch(e) {
+        reject(e);
```
What error would we expect to happen here?

stefanpenner replied:
none, this is to handle the exceptional cases.

stefanpenner replied:
This also maintains existing behavior. Currently if any of those lines error, the promise constructor will catch them.  As this PR moves them to be async, we must manage that ourselves.

trentmwillis commented on addon/loaders/js.js:
```diff
@@ -15,11 +15,17 @@ export default nodeLoader(function js(uri) {
     }
 
-    const script = createLoadElement('script', resolve, reject);
+    Ember.run.schedule('afterRender', () => {
```
We should add a comment as to why this `afterRender` is present.

trentmwillis reviewed (changes requested):
I wouldn't say it is "kinda strange", it is just the simple approach. The current approach doesn't cause any abnormal behavior, but it is likely non-optimal.

Should we do the same thing for CSS loader? It likely isn't as big of a deal as loading JS mid-run-loop, but seems like it may still be beneficial.

Commit:
Fixup test assertions.

We cannot simply return the promise in the failure scenarios (since
returning a rejected promise will fail the test).  I have added
back a small assertion confirming the rejections in cases we _know_
should reject.

The CSS load failure scenario is a bit annoying/odd because we cannot
guarantee that we fail properly when the asset cannot be loaded (due
to lack of browser support for `onload`/`onerror` callbacks with the
`<link>` element).

rwjblue commented:
Rebased this on top of latest master, and fixed the two failing assertions (both around `service.loadAsset()` rejecting). 

@stefanpenner  / @trentmwillis - This is ready for another look.

trentmwillis commented on addon/loaders/css.js:
```diff
@@ -19,34 +20,46 @@ export default nodeLoader(function css(uri) {
     }
 
-    // Try using the default onload/onerror handlers...
-    const link = createLoadElement('link', resolve, reject);
```
Why move this declaration outside the block where it is initialized?

trentmwillis reviewed (approved):
One small nitpick, but otherwise looks good to me!

rwjblue commented:
@stefanpenner - Mind rebasing?

villander commented:
@stefanpenner mind rebasing? please

stefanpenner commented:
sure

villander commented:
@stefanpenner I'd love to see this merge, you want some help?

stefanpenner commented:
@villander yes please

villander commented:
@stefanpenner conflicts solved by #81
